packages/cf-util-http/src/http.js

```diff
@@ -33,84 +25,136 @@ export function beforeSend(callback) {
  * });
  * ```
  *
  * @param {String} method - GET/POST/PUT/PATCH/DELETE
  * @param {String} url
- * @param {Object} [opts]
- * @param {Object} [opts.parameters]
- * @param {Object} [opts.headers]
- * @param {Object} [opts.body]
- * @param {Function} [callback]
- * @returns {Function} Abort request.
+ * @param {Object} [opts] Any options fetch()'s second parameter accepts
+ * @param {Object|Array<Array>|Map} [opts.parameters] Query parameters to add to `url`
+ * @returns {Promise} Fetch promise
+ *
+ * @see [Global Fetch on MDN](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)
+ * @see [WHATWG Fetch API Spec](https://fetch.spec.whatwg.org/)
  */
-export function request(method, url, opts, callback) {
-  opts = opts || {};
-
-  opts.method = method;
+export function request(method, url, opts = {}, callback) {
+  opts.method = method.toUpperCase();
   opts.url = url;
   opts.callback = callback;
 
   // Allow beforeSend to modify request options.
   beforeSendCallbacks.forEach(cb => cb(opts));
 
-  // Configure request
-  const req = superagent[METHODS[opts.method]](opts.url);
+  method = opts.method;
+  url = opts.url;
+  callback = opts.callback;
 
   if (opts.parameters) {
-    req.query(opts.parameters);
-  }
-
-  if (opts.headers) {
-    req.set(opts.headers);
-  }
-
-  if (opts.body) {
-    req.send(opts.body);
+    if (url.indexOf('?') === -1) {
+      url = `${url}?${toQueryParams(opts.parameters)}`;
+    }
   }
 
-  let logMessage = `${opts.method} ${opts.url}`;
+  let logMessage = `${method} ${url}`;
 
   logRequest(logMessage);
 
-  // Send request
-  req.end((err, res) => {
-    logMessage = `${logMessage} (${res.status} ${res.statusText})`;
-
-    const result = {
-      headers: res.headers,
-      status: res.status,
-      body: res.body,
-      text: res.text
-    };
+  return fetch(url, opts)
+    .then(response => {
+      const headers = toHash(response.headers);
+      const status = response.status;
+      logMessage = `${logMessage} (${status} ${response.statusText})`;
+
+      if (response.ok) {
+        logSuccess(logMessage);
+
+        return response.text().then(text => {
+          if (callback) {
+            callback(
+              undefined,
+              wrapResponse(headers, status, text, text, response)
+            );
+          }
+          return response;
+        });
+      }
 
-    if (err) {
       logError(logMessage);
-      opts.callback(result);
-    } else {
-      logSuccess(logMessage);
-      opts.callback(null, result);
-    }
-  });
-
-  // Return abort function
-  return function() {
-    req.abort();
-  };
+      return response.text().then(text => {
+        if (callback) {
+          callback(wrapResponse(headers, status, text, text, response));
+        }
+        return response;
+      });
+    })
+    .catch(console.trace); // Unrecoverable errors
```

Comment:
The promise is returned, so the caller can catch it there and do the appropriate thing. The error is unrecoverable as this point by fetch, I just we should just log it and let the caller handle it, or ignored if desired. The only `error` the caller can handle is anything not 2xx. The callback is called with an error in a resolved promise state to preserve the existing API.